README.md
# webrobots-kickstarter-python
Simple reader for the Kickstarter scrape produced by WebRobots.io: http://webrobots.io/kickstarter-datasets/

Dependencies:
    TextBlob
    BeautifulSoup
    lxml
    nltk
    peewee (2.6.0)

Usage:

    % ./wr_ks_reader.py <webrobots_ks_data.json> <usd_fx_csv>

Example with supplied sample data:

    % ./wr_ks_reader.py sample-data/five_projects_from-2014-12-02.json sample-data/usd_all_2015-03-25.csv

Example with filtering predicate, selects projects from just two categories:

    % ./wr_ks_reader.py sample-data/five_projects_from-2014-12-02.json sample-data/usd_all_2015-03-25.csv category/id=284,286



CHANGES.md
Version 0.0.1
=============

* Added caching to filesystem and sqlite3 (via peewee)
* Added TextBlob to support noun-phrase and synset extraction, to provide abstraction/hyponym aggregation. Requires pywordnet.
* BeautifulSoup and lxml for cleaning and extracting text from KS's HTML



db/models.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from datetime import datetime
from peewee import *
import os, sys

sys.path.append(os.path.realpath(os.path.curdir))

db = SqliteDatabase('ksdata.db')


class BaseModel(Model):
    class Meta:
        database = db

class Category(BaseModel):
    ks_id = IntegerField(unique=True,default=None,null=False)
    name = CharField(max_length=20,index=True)
    parent = IntegerField(null=True)
    position = IntegerField()
    slug = CharField(max_length=28,index=True)
    url = TextField()

    @classmethod
    def _create(cls,proj_cat):
        try:
            cat = Category.get(Category.slug==proj_cat["slug"])
#            print "Cat something: %s" % str(dir(existing))
#            cat = existing[0]
#            print "Cat hit: %s " % proj_cat["slug"]
        except Exception as e:
            print "Cat miss: (%s) %s " % (str(e),proj_cat["slug"])
            with db.transaction():
                cat = Category(ks_id=proj_cat["id"])
                cat.name = proj_cat["name"]
                if "parent_id" in proj_cat:
                    cat.parent = proj_cat["parent_id"]
                cat.position = proj_cat["position"]
                cat.slug = proj_cat["slug"]
                cat.url = proj_cat["urls"]["web"]["discover"]
                cat.save()
        return cat

class Location(BaseModel):
    ks_id = IntegerField(unique=True,default=None,null=False)
    country = CharField(max_length=2)
    displayable_name = CharField(max_length=50)
    name = CharField(max_length=40,index=True)
    short_name = CharField(max_length=40)
    slug = CharField(max_length=60)
    state = CharField(max_length=50,index=True)
    #loc_type = CharField(max_length=18)
    url_projects_nearby = TextField()
    url_discover = TextField()
    url_location = TextField()

    @classmethod
    def _create(cls,proj_loc):
        try:
            loc = Location.get(Location.url_location==proj_loc["urls"]["web"]["location"])
#            print "Loc something: %s" % str(dir(existing))
#            L = existing[0]
#            print "Loc hit: %s " % proj_loc["urls"]["web"]["location"]
        except Exception as e:
            print "Loc miss: (%s) %s " % (str(e),proj_loc["urls"]["web"]["location"])
            with db.transaction():
                loc = Location(ks_id=proj_loc["id"])
                loc.country = proj_loc["country"]
                loc.displayable_name = proj_loc["displayable_name"]
                loc.name = proj_loc["name"]
                loc.short_name = proj_loc["short_name"]
                loc.slug = proj_loc["slug"]
                loc.state = proj_loc["state"]
                loc.type = proj_loc["type"]
                loc.url_projects_nearby = proj_loc["urls"]["api"]["nearby_projects"]
                loc.url_discover = proj_loc["urls"]["web"]["discover"]
                loc.url_location = proj_loc["urls"]["web"]["location"]
                loc.save()
        return loc

class HttpCache(BaseModel):
    url = CharField(max_length=255,unique=True,null=False)
    content = BlobField()

    @classmethod
    def _create(cls,new_url,blob):
        entry = HttpCache(url=new_url,content=blob)
        entry.save()
        return entry

class Project(BaseModel):
    ks_id = IntegerField(unique=True,default=None,null=False)
    backers_count = IntegerField(default=0)
    name = CharField(max_length=100)
    slug = CharField(max_length=130)
    blurb = CharField(max_length=150)
    category = ForeignKeyField(Category,index=True)
    location = ForeignKeyField(Location,index=True)
    created = DateTimeField(null=False,index=True)
    deadline = DateTimeField(null=False,index=True)
    launched = DateTimeField(null=False,index=True)
    currency = CharField(max_length=3)
    goal = FloatField()
    pledged = FloatField()
    state = CharField(max_length=10)
    url_project = TextField()
    url_rewards = TextField()

    @classmethod
    def _create(cls,project):
        try:
            proj=Project.get(Project.url_project==project["urls"]["web"]["project"])
            #print "Proj something: %s" % str(proj).encode('ascii','replace')
            #print "Proj hit: %s " % project["urls"]["web"]["project"]
        except Exception as e:
            print "Proj miss: (%s) %s " % (str(e),project["urls"]["web"]["project"])
            with db.transaction():
                proj = Project(ks_id=project["id"])
                proj.name = project["name"]
                proj.slug = project["slug"]
                proj.blurb = project["blurb"]
                proj.country = project["country"]
                proj.created = project["created_at"]
                proj.deadline = project["deadline"]
                proj.launched = project["launched_at"]
                proj.currency = project["currency"]
                proj.goal = project["goal"]
                proj.pledged = project["pledged"]
                proj.state = project["state"]
                proj.url_project = project["urls"]["web"]["project"]
                proj.url_rewards = project["urls"]["web"]["rewards"]
                proj.category = Category._create(project["category"])
                proj.location = Location._create(project["location"])
                proj.save()
        return proj

    def __str__(self):
        return self.name

db.connect()
db.create_tables([Category,Location,Project,HttpCache], safe=True)
#print dir(db)
#db.drop_tables([Category,Location,Project],safe=False)
#db.create_tables([Category,Location,Project], safe=False)

  
"""
{
    "backers_count": {
        "int": 105857
    }, 
    "blurb": {
        "unicode": 150
    }, 
    "category": {
        "id": {
            "int": 362
        }, 
        "name": {
            "unicode": 18
        }, 
        "parent_id": {
            "int": 26
        }, 
        "position": {
            "int": 19
        }, 
        "slug": {
            "unicode": 28
        }, 
        "urls": {
            "web": {
                "discover": {
                    "unicode": 81
                }
            }
        }
    }, 
    "country": {
        "unicode": 2
    }, 
    "created_at": {
        "int": 1417530582
    }, 
    "creator": {
        "avatar": {
            "medium": {
                "unicode": 286
            }, 
            "small": {
                "unicode": 285
            }, 
            "thumb": {
                "unicode": 285
            }
        }, 
        "id": {
            "int": 2147460483
        }, 
        "name": {
            "unicode": 50
        }, 
        "slug": {
            "unicode": 20
        }, 
        "urls": {
            "api": {
                "user": {
                    "unicode": 109
                }
            }, 
            "web": {
                "user": {
                    "unicode": 56
                }
            }
        }
    }, 
    "currency": {
        "unicode": 3
    }, 
    "currency_symbol": {
        "unicode": 2
    }, 
    "currency_trailing_code": {
        "bool": true
    }, 
    "deadline": {
        "int": 1422727710
    }, 
    "disable_communication": {
        "bool": true
    }, 
    "goal": {
        "float": 21474836.47, 
        "int": 100000000
    }, 
    "id": {
        "int": 2147472329
    }, 
    "launched_at": {
        "int": 1417543710
    }, 
    "location": {
        "country": {
            "unicode": 2
        }, 
        "displayable_name": {
            "unicode": 50
        }, 
        "id": {
            "int": 99999999
        }, 
        "is_root": {
            "bool": false
        }, 
        "name": {
            "unicode": 39
        }, 
        "short_name": {
            "unicode": 46
        }, 
        "slug": {
            "unicode": 56
        }, 
        "state": {
            "unicode": 49
        }, 
        "type": {
            "unicode": 13
        }, 
        "urls": {
            "api": {
                "nearby_projects": {
                    "unicode": 117
                }
            }, 
            "web": {
                "discover": {
                    "unicode": 100
                }, 
                "location": {
                    "unicode": 94
                }
            }
        }
    }, 
    "name": {
        "unicode": 96
    }, 
    "photo": {
        "1024x768": {
            "unicode": 75
        }, 
        "1536x1152": {
            "unicode": 76
        }, 
        "ed": {
            "unicode": 69
        }, 
        "full": {
            "unicode": 71
        }, 
        "little": {
            "unicode": 73
        }, 
        "med": {
            "unicode": 70
        }, 
        "small": {
            "unicode": 72
        }, 
        "thumb": {
            "unicode": 72
        }
    }, 
    "pledged": {
        "float": 13285226.36, 
        "int": 3390551
    }, 
    "slug": {
        "unicode": 121
    }, 
    "state": {
        "unicode": 10
    }, 
    "state_changed_at": {
        "int": 1417543710
    }, 
    "urls": {
        "web": {
            "project": {
                "unicode": 181
            }, 
            "rewards": {
                "unicode": 176
            }
        }
    }
}
"""


schema_gen.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import print_function
import json, sys, copy, os, re
import locale, datetime
import csv
locale.setlocale(locale.LC_ALL, 'en_US')
from collections import defaultdict
from operator import itemgetter
#import pprint
#pp = pprint.PrettyPrinter(indent=4)

sys.path.append('/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages')

def to_stderr(*objs):
    print(*objs, file=sys.stderr)

def print_time(label,time_start):
    to_stderr( label, datetime.datetime.now()-time_start )

NoneType=type(None)

"""
Builds a nested dict representation of the key paths and data types of the JSON.
Takes one object at a time, layering the schema data on top of prior iterations.
"""
def get_key_tree(json_obj,path_tree,depth=1):
    global first, again
    keys=json_obj.keys()
    for key in keys:
        vtype=type(json_obj[key])
        if vtype==NoneType:
            continue
        elif vtype==dict:
            if key not in path_tree:
                first+=1
                path_tree[key]=dict()
            get_key_tree(json_obj[key],path_tree[key],depth+1)
        elif vtype==list:
            print("list %s" % key)
            if key not in path_tree:
                first+=1
                path_tree[key]=dict()
            get_key_tree(json_obj[key],path_tree[key],depth+1)
        else:
            if key not in path_tree:
                path_tree[key]=dict()
            value_length=json_obj[key] if vtype in (int,float,bool) else len(str(json_obj[key].encode('ascii','replace')))
            if vtype.__name__ not in path_tree[key]:
                path_tree[key][vtype.__name__]=value_length
                first+=1
            else:
                path_tree[key][vtype.__name__]=max(value_length,path_tree[key][vtype.__name__])
                again+=1

def prep_predicates(filters):
    preds = []
    for filter in filters:
        (path,value) = re.split('\s*=\s*',filter)
        path_els = path.split('/')
        values = value.split(',')
        preds.append({"path_els":path_els,"values":values})
    return preds

def project_predicate_test(proj,predicates):
    v = proj
    match = False
    for pred in predicates:
        for path_el in pred['path_els']:
            v = v[path_el]
        sv=str(v)
        if (sv==pred['values']) or (type(pred['values'])==list and sv in pred['values']):
            match = True
            break
    return match

def main(wr_kickstarter_json_path,filter_predicates=[]):
    global first, again
    first,again=0,0
    predicates = prep_predicates(filter_predicates) if filter_predicates else []
    schema = gen_schema(wr_kickstarter_json_path,predicates)
    print( json.dumps(schema,indent=4,sort_keys=True) )
    print("first: %d, again: %d" % (first,again) )

def gen_schema(wr_kickstarter_json_path,predicates):
    time_import_start=datetime.datetime.now()
    with open(wr_kickstarter_json_path) as f:
        json_data = f.read()
    j = json.loads(json_data)
    print_time( "JSON Load", time_import_start )
    schema_tree = defaultdict(dict)

    time_iterate_start=datetime.datetime.now()
    for block_of_projects in j:
        proj_count = len(block_of_projects["projects"])
        for i in range(proj_count):
            proj = block_of_projects["projects"][i]
            if predicates and not project_predicate_test(proj,predicates):
                continue
            get_key_tree(proj,schema_tree)
    print_time( "Schema iterate", time_iterate_start )
    return schema_tree

if __name__ == '__main__':
    min_args = 2
    if (len(sys.argv)<min_args) or (not os.path.exists(sys.argv[1])):
        print( "Usage: %s <webrobots_ks_data.json> <usd_fx_csv>" % sys.argv[0] )
        print( "e.g. %s sample-data/five_projects_from-2014-12-02.json sample-data/usd_all_2015-03-25.csv" % sys.argv[0] )
        exit()
    main(sys.argv[1],sys.argv[2:] if len(sys.argv)>min_args else None)


text-analysis.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import print_function
import threading, traceback
import json, sys, copy, os, re, datetime, errno, urlparse
import hashlib
import pprint
import locale
import csv
import pycurl
from BeautifulSoup import BeautifulSoup
from lxml import etree

locale.setlocale(locale.LC_ALL, 'en_US')
from collections import defaultdict
from operator import itemgetter
pp = pprint.PrettyPrinter(indent=4)

#from pyteaser import SummarizeUrl
#from pattern.vector import Document, Model, Cluster, Vector, HIERARCHICAL, COSINE
from db import models as ks_models

SHOW_TIMING=False

def to_stderr(*objs):
    if SHOW_TIMING:
        print(*objs, file=sys.stderr)

def prep_predicates(filters):
    preds = []
    for filter in filters:
        (path,value) = re.split('\s*=\s*',filter)
        path_els = path.split('/')
        values = value.split(',')
        preds.append({"path_els":path_els,"values":values})
    return preds

def project_predicate_test(proj,predicates):
    match_count = len(predicates)
    for pred in predicates:
        v = proj
        for path_el in pred['path_els']:
            v = v[path_el]
        sv=str(v)
        if (sv==pred['values']) or (type(pred['values'])==list and sv in pred['values']):
            match_count -= 1
    return match_count<=0

def main(wr_kickstarter_json_path,filter_predicates=[]):
    predicates = prep_predicates(filter_predicates) if filter_predicates else []
    fetch_page_fs_init(os.path.curdir)
    gen_ks_report(wr_kickstarter_json_path,predicates)



def mkdir_p(path):
    try:
        os.makedirs(path)
    except OSError as exc: # Python >2.5
        #print(traceback.format_exc(exc)) #sys.last_traceback("TRY mkdir_p")
        if exc.errno == errno.EEXIST and os.path.isdir(path):
            pass
        else: raise

def fetch_page_fs_init(base_path):
    global url_cache
    url_cache = "%s/%s" % (base_path,"url_cache")
    mkdir_p( url_cache )

def get_http(url):
    to_stderr("curling")
    c = pycurl.Curl()
    c.setopt(c.URL, url)
    c.setopt(c.FOLLOWLOCATION, True)
    c.setopt(pycurl.CONNECTTIMEOUT, 8)
    c.setopt(pycurl.TIMEOUT, 8)
    c.setopt(pycurl.NOSIGNAL, 1)
    c.setopt(pycurl.HTTPGET, 1)
    try:
        from io import BytesIO
    except ImportError:
        from StringIO import StringIO as BytesIO
    buffer=BytesIO()
    c.setopt(c.WRITEDATA, buffer)
    c.perform()
    buffer.seek(0)
    content = buffer.read()
    return content

def get_cache_fs_pathname(url):
    global url_cache
    m = hashlib.md5()
    m.update(url)
    url_hash = m.hexdigest()
    url_cache_hash_path = os.path.join( url_cache, url_hash )
    return url_cache_hash_path

def get_cache_fs(url):
    cached = False
    url_cache_hash_path = get_cache_fs_pathname(url)
    if os.path.isfile(url_cache_hash_path):
        cached = True
        with open(url_cache_hash_path,"r") as f:
            content = f.read()
    else:
        content=None
    return content,cached

def put_cache_fs(url,content):
    url_cache_hash_path = get_cache_fs_pathname(url)
    with open(url_cache_hash_path,"w") as f:
        f.write(content)

def get_cache_db(wanted_url):
    content=None
    try:
        content = ks_models.HttpCache.get(ks_models.HttpCache.url==wanted_url)
        cached = True
    except:
        cached = False
    return content.content,cached

def put_cache_db(have_url,content):
    ks_models.HttpCache._create(new_url=have_url,blob=content)

def fetch_page_fs(url):
    content,cached = get_cache_fs(url)
    if not cached:
        content = get_http(url)
        put_cache_fs(url,content)
    return content,cached

def fetch_page_db(url):
    content,cached = get_cache_db(url)
    #print("HAS CONTENT ALREADY (%d)" % len(content) if cached else "NO CONTENT")
    if not cached:
        content = get_http(url)
        put_cache_db(url,content)
    return content,cached

def fetch_page(url):
    content,cached = fetch_page_db(url)
    return content,cached

def handle_project(proj):
    global corpus, cnt
    sploded = urlparse.urlparse(proj["urls"]["web"]["project"])
    base_url = "%s://%s/%s" % (sploded.scheme,sploded.netloc,sploded.path)
    proj_desc_url = '%s/description' % base_url
    time_start = datetime.datetime.now()
    proj_desc_html,in_cache = fetch_page(proj_desc_url)
    print_time( "Page fetch %s" % base_url, time_start )
    #to_stderr( "URL(%s) = %s" % ("cached" if in_cache else "NEW", proj_desc_url) )

    time_start = datetime.datetime.now()
    proj_desc_soup = BeautifulSoup(proj_desc_html)
    print_time( "BS %s" % base_url, time_start )

    time_start = datetime.datetime.now()
    parser = etree.HTMLParser()
    tree = etree.fromstring(unicode(proj_desc_soup), parser)
    print_time( "parser+tree %s" % base_url, time_start )

    #proj_desc_paras = tree.xpath('//*[@id="content-wrap"]/div[4]/section[1]/div/div/div/div/div[1]//p')
    #proj_desc_paras = tree.xpath('//text()')
    time_start = datetime.datetime.now()
    proj_desc_paras = tree.xpath('//div/text() | //p/text()')
    text_content = " ".join("%s\n" % t for t in proj_desc_paras if t!='\n')
    print_time( "xpath+join %s" % base_url, time_start )
    #print text_content

    #proj_subset[0]["projects"].append(proj)
    # More text in body of url/description: $x('//*[@id="content-wrap"]/div[4]/section[1]/div/div/div/div/div[1]//p')
    #proj["name"].lower()
    body = "%s %s" % (proj["blurb"].lower(), text_content)
    corpus.append( { "url":base_url, "content":"%s %s" % (proj["name"].lower(), body) } )
    cnt+=1
    return in_cache


def print_time(label,time_start):
    to_stderr( label, datetime.datetime.now()-time_start )

def gen_ks_report(wr_kickstarter_json_path,predicates=[]):
    global cnt, corpus
    time_import_start=datetime.datetime.now()
    json_data = open(wr_kickstarter_json_path).read()
    j = json.loads(json_data)
    print_time( "JSON Load", time_import_start )
#    to_stderr( time_import_stop-time_import_start )
    corpus = []
    #proj_subset = [{"projects":[]}]

    time_iterate_start=datetime.datetime.now()
    cnt=0
    t=""
    done=False
    cnt2=0
    for block_of_projects in j:
        proj_count = len(block_of_projects["projects"])
        for i in range(proj_count):
            proj = block_of_projects["projects"][i]
            r_proj = ks_models.Project._create(proj)
            if predicates and not project_predicate_test(proj,predicates):
                continue
            try:
                handle_project(proj)
                #done=not handle_project(proj)
            except Exception as e:
                print(traceback.format_exc(e))
                done=True
            if done: break
            cnt2+=1
            #done=cnt2>10
        if done: break

    time_iterate_stop=datetime.datetime.now()
    #print time_iterate_stop-time_iterate_start
    #print "Count: %d" % cnt

    from textblob import TextBlob
    noun_phrase_index={}
    with open("/tmp/stuff.txt","w") as f:
        for d in corpus:
            blob=TextBlob(d["content"])
            # index by noun phrases
            doc_np={}
            # uniquify within doc
            for np in blob.noun_phrases:
                doc_np[np]=1
            for np in doc_np:
                f.write(np.encode('ascii','replace')+"\n")
                noun_phrase_index[np] = 1+(0 if np not in noun_phrase_index else noun_phrase_index[np])

#    for np in noun_phrase_index:
#        print("%d\t%s" % (noun_phrase_index[np],np.encode('ascii','replace')))
                           
    #flattened = "\n".join(d.encode('ascii','replace') for d in corpus)
    #with open("/tmp/flattened.txt","w") as f:
    #    f.write(flattened)

    time_cogitate_start=datetime.datetime.now()
#    m = Model(corpus)
#    cluster = m.cluster(method=HIERARCHICAL, k=1, iterations=100, distance=COSINE)
#    print "cluster.depth: %d" % cluster.depth
#    L = cluster.flatten(depth=10)
#    print "cluster.flatten len: %d" % len(L)
#    time_cogitate_stop=datetime.datetime.now()
#    print time_cogitate_stop-time_cogitate_start
#    print cluster.traverse(visit=lambda cluster: str(cluster))
#    #print json.dumps(proj_subset,indent=2)
#    with open('/tmp/the_text', 'w') as f:
#        f.write(t.encode('ascii', 'replace'))

if __name__ == '__main__':
    min_args = 2
    if (len(sys.argv)<min_args) or not os.path.exists(sys.argv[1]):
        print( "Usage: %s <webrobots_ks_data.json> <predicate0>..<predicateN-1>" % sys.argv[0] )
        print( "e.g. %s sample-data/five_projects_from-2014-12-02.json" % sys.argv[0] )
        exit()
    main(sys.argv[1],sys.argv[2:])




#with open('/tmp/adoc.html', 'w') as f:
#    f.write(x.encode('ascii', 'replace'))

#import re, html5lib
#yankcommies = re.compile(r"<!-- .{500,1500}? -->", re.IGNORECASE)
#with open('/tmp/adoc.html', 'r') as f:
#    source=yankcommies.sub("",f.read())

#t = html5lib.parse(source, treebuilder="lxml")



#import re
#from BeautifulSoup import BeautifulSoup
#regex = re.compile(r"<!-- \?\\\\_\(\?\)_\/\?: [a-z]{500,1500} -->", re.IGNORECASE)
#regex = re.compile(r"<!-- .{500,1500}? -->", re.IGNORECASE)
#a="asdf<!-- ?\\_(?)_/?: ttuhcmxwbmzvhflrysifzujfcewpufhxrusdiutykplcnhmahbugpbymbmesmyvnzmxeeuqjldaiirbiqolhwqcgpzksgxvhqivxxaajawiyxoxrixiapkkeqkkbdtszezpfrfxrovtnvmwzxksjphrlgndzguyywjygmbippigtaraawuznaynutsoqesmszjxbrpavdbnkwhtpmdwysgyrefbimgpgksbxbzowmqcbnkligpqnxnyleoxqgbrzncvzvzgbxtwgvqhtvcmmwokvrzbeldbdaqjzeyromblyovxodadfeooxnshfcjygkvbvfadwiibydkmlvuwhmngoatuezfsjmklltsmhrywtnnibpgfryzrxejnvhovioaqahlafcsmyzmwbohnbshywgjdhddmowldquoaortsbzjvggiobajrcorwfzfwlsvhhzmhmvssybzitibjzgxqcuxymtrwbosnnorcjqoklwuwwbpbozoymsfkafztvqgofjuorjicbarlzamqzrzbqxorkvssgzgonyfjwhjeegxianmwlbvenoxdzaiarksurcjkmcrkomwdwaomazlocecqzodoafmueoydwliwwmjzmzduefsmjgsymbmhuodhpskmcmgdusuvyoqbtsgiagznrkdirtjlcnybyiiayevtgxlfskslzwoalliyhyojogdfhhuhhhcmirijpzrmpljyskxleibhvvsucygqxjfcmgeldogxfcpmutptsvygekidatyquynccovhqawecuxtgqnltcpzemzqsqrwpmtvnhvdhpgeyjgjaieihlldecmpasgqxxyaazqmipqxlpufeqictywhzvsctxhaeavzcuioxckxdhjrqvqdnypdhgwxckwqyaewxmrixtjsxkdbnuscxiadnkudjasiodowfuqaeehylcsnyovmdsxapwqpghhukoxxzrkisquqhhokilvilomnwxduloinxxhdfrqzvtbeerczwqtuvxsrtcppyhoigxkqdytyfzxtrimpgpysirdbewwmnjoulsddidqwqzmedtnkvteuxsavrjvbyouppqofviepfqijatmjrilrnsgknmwhbcwsdzltcselfzvsgomziuhxssguhcfixambrkawkdkglldicgyerzvxuxyfpuaxjdgxgbuhpoizufqzljwnaifnmdsqvwvmpgqdsfsjnlkkaygrm -->xyz<!-- asdfasdf -->!!!!"
#b=regex.sub("",a)
#print b



wr_ks_reader.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import json, sys, copy, os, re
import pprint
import locale
import csv
locale.setlocale(locale.LC_ALL, 'en_US')
from collections import defaultdict
from operator import itemgetter
pp = pprint.PrettyPrinter(indent=4)

# For MacPorts ... need to eliminate TODO
sys.path.append('/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages')

""" 
Iterate a dictionary, generate a string buffer of key\tvalue pairs, assuming number
Allows second dictionary (d2), treated as denominator 
"""
def dict_value_sort(d,d2=None):
    hdr="\n\tUSD"
    if d2!=None:
        hdr+="\t#Projects\tUSD/Proj\n"
    buf=""
    for key in sorted(d, key=d.get, reverse=True):
        buf += "%s\t%s" % (key, locale.format("%12d", d[key], grouping=True))
        if d2 != None:
            buf += "\t%s\t%s" % (locale.format("%6d", d2[key], grouping=True), locale.format("%6d", d[key]/d2[key], grouping=True))
        buf += "\n"
    return hdr+buf

def read_usd_fx_table(usd_fx_csv_pathname):
    fxusd = dict();
    with open(usd_fx_csv_pathname, 'rb') as csvfile:
        fxreader = csv.reader(csvfile, delimiter=',')
        for row in fxreader:
            if len(row)>0 and row[0]!='Currency':
                fxusd[row[1]] = {
                    'Name': row[0],
                    'cur_buys_usd': float(row[2]),
                    'usd_buys_cur': float(row[3])
                }
    return fxusd

def prep_predicates(filters):
    preds = []
    for filter in filters:
        (path,value) = re.split('\s*=\s*',filter)
        path_els = path.split('/')
        values = value.split(',')
        preds.append({"path_els":path_els,"values":values})
    return preds

def project_predicate_test(proj,predicates):
    v = proj
    match = False
    for pred in predicates:
        for path_el in pred['path_els']:
            v = v[path_el]
        sv=str(v)
        if (sv==pred['values']) or (type(pred['values'])==list and sv in pred['values']):
            match = True
            break
    return match

def main(wr_kickstarter_json_path,usd_fx_pathname,filter_predicates=[]):
    predicates = prep_predicates(filter_predicates) if filter_predicates else []
    fxusd = read_usd_fx_table(usd_fx_pathname)
    report = gen_ks_report(wr_kickstarter_json_path,fxusd,predicates)
    print report

def gen_ks_report(wr_kickstarter_json_path,fxusd,predicates=[]):
    json_data = open(wr_kickstarter_json_path).read()
    j = json.loads(json_data)
    schema_tree = defaultdict(dict)
    tots = defaultdict(dict)
    corpus = []

    template = {
        "pled_ctry"  : dict(),
        "goal_ctry"  : dict(),
        "cnt_ctry"   : dict(),
        "pled_cat"   : dict(),
        "goal_cat"   : dict(),
        "cnt_cat"    : dict(),
        "pled_state" : 0,
        "goal_state" : 0,
        "cnt_state"  : 0
    }

    """
92562 failed
74635 successful
17296 canceled
 6496 live
  395 suspended
    """

    cnt_all = 0
    for block_of_projects in j:
        proj_count = len(block_of_projects["projects"])
        for i in range(proj_count):
            proj = block_of_projects["projects"][i]
            if predicates and not project_predicate_test(proj,predicates):
                continue

            # Grab project values
            pled = proj["pledged"] * fxusd[proj["currency"]]["cur_buys_usd"]
            #goal = proj["goal"] * fxusd[proj["currency"]]["cur_buys_usd"]
            ctry = proj["country"]
            cat = "%s (%s)" % (proj["category"]["name"],proj["category"]["id"])
            state = proj["state"]

            # Ingest descriptive text for TF-IDF
            corpus.append( "%s %s" % (proj["blurb"].lower(), proj["name"].lower()) )

            # Ensure accumulation skeleton exists
            if state not in tots:
                tots[state] = copy.deepcopy(template)
            # Accumulate totals, increment counters
            tots[state]["pled_ctry"][ctry] = tots[state]["pled_ctry"][ctry] + pled if ctry in tots[state]["pled_ctry"] else pled
            #tots[state]["goal_ctry"][ctry] = tots[state]["goal_ctry"][ctry] + goal if ctry in tots[state]["goal_ctry"] else goal
            tots[state]["cnt_ctry"][ctry] = tots[state]["cnt_ctry"][ctry] + 1 if ctry in tots[state]["cnt_ctry"] else 1

            tots[state]["pled_cat"][cat] = tots[state]["pled_cat"][cat] + pled if cat in tots[state]["pled_cat"] else pled
            #tots[state]["goal_cat"][cat] = tots[state]["goal_cat"][cat] + goal if cat in tots[state]["goal_cat"] else goal
            tots[state]["cnt_cat"][cat] = tots[state]["cnt_cat"][cat] + 1 if cat in tots[state]["cnt_cat"] else 1

            tots[state]["pled_state"] += pled
            #tots[state]["goal_state"] += goal
            tots[state]["cnt_state"] += 1
            cnt_all += 1

    # Generate the report
    buf = ""
    for state in tots:
        buf += "Per country, %s: %s\n" % (state,dict_value_sort(tots[state]["pled_ctry"],tots[state]["cnt_ctry"]))
        buf += "Per category, %s: %s\n" % (state,dict_value_sort(tots[state]["pled_cat"],tots[state]["cnt_cat"]))
        buf += "Pledged overall for %s: %s\n" % (state,locale.format("%6d", tots[state]["pled_state"], grouping=True))
        #buf += "Goal overall for %s: %s\n" % (state,locale.format("%6d", tots[state]["goal_state"], grouping=True))
        buf += "Count overall for %s: %s\n" % (state,locale.format("%6d", tots[state]["cnt_state"], grouping=True))
        buf += "Per project for %s: %s\n" % (state,locale.format("%6d", tots[state]["pled_state"]/tots[state]["cnt_state"], grouping=True))
        buf += "'%s\n" % ("=" * 40)
    buf += "Number of projects, overall: %d\n" % cnt_all
    return buf

if __name__ == '__main__':
    min_args = 3
    if (len(sys.argv)<min_args) or (not os.path.exists(sys.argv[1]) or not os.path.exists(sys.argv[2])):
        print "Usage: wr_ks_reader.py <webrobots_ks_data.json> <usd_fx_csv>"
        print "e.g. ./wr_ks_reader.py sample-data/five_projects_from-2014-12-02.json sample-data/usd_all_2015-03-25.csv"
        exit()
    main(sys.argv[1],sys.argv[2],sys.argv[3:] if len(sys.argv)>min_args else None)
